Source/WebGPU/WebGPU/BindGroup.mm

```diff
@@ -58,14 +87,18 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
     Vector<BindableResource> resources;
 #if HAVE(TIER2_ARGUMENT_BUFFERS)
     if ([m_device argumentBuffersSupport] != MTLArgumentBuffersTier1) {
-        const BindGroupLayout& bindGroupLayout = WebGPU::fromAPI(descriptor.layout);
-
-        auto vertexArgumentBuffer = bindGroupLayout.vertexArgumentBuffer();
-        auto fragmentArgumentBuffer = bindGroupLayout.fragmentArgumentBuffer();
-        auto computeArgumentBuffer = bindGroupLayout.computeArgumentBuffer();
-        char* argumentBufferContents[3] = { (char*)vertexArgumentBuffer.contents, (char*)fragmentArgumentBuffer.contents, (char*)vertexArgumentBuffer.contents };
-        size_t bindingForStage[] = { 0, 0, 0 };
-        for (uint32_t i = 0; i < descriptor.entryCount; ++i) {
+        uint32_t entryCount = descriptor.entryCount;
+        auto bindGroupLayout = descriptor.layout ? &WebGPU::fromAPI(descriptor.layout) : nullptr;
+        auto bufferSizes = sizeOfEntries(descriptor, bindGroupLayout);
+        constexpr WGPUShaderStage shaderStage[] = { WGPUShaderStage_Vertex, WGPUShaderStage_Fragment, WGPUShaderStage_Compute };
+        constexpr auto shaderStageLength = WTF_ARRAY_LENGTH(shaderStage);
+        id<MTLBuffer> argumentBuffer[shaderStageLength];
+        for (size_t i = 0; i < shaderStageLength; ++i)
+            argumentBuffer[i] = bufferSizes[i] ? safeCreateBuffer(bufferSizes[i], MTLStorageModeShared) : nil;
```

Comment:
I still think we shouldn’t be manually calculating byte sizes ourself but should be using MTLArgumentEncoder.encodedLength